### PR TITLE
azure-functions-core-tools: 4.0.5455 -> 4.0.6594

### LINKED ISSUE #360154: Build failure: azure-functions-core-tools
## Steps To Reproduce

1. See under context

## Build log

<details>

<summary>Build Log</summary>

```
error:
       … while calling the 'derivationStrict' builtin
         at <nix/derivation-internal.nix>:9:12:
            8|
            9|   strict = derivationStrict drvAttrs;
             |            ^
           10|

       … while evaluating derivation 'MitC'
         whose name attribute is located at /nix/store/k5g5frkf231kzsrnc8szlg229ychpcqy-nixpkgs/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:336:7

       … while evaluating attribute 'nativeBuildInputs' of derivation 'MitC'
         at /nix/store/k5g5frkf231kzsrnc8szlg229ychpcqy-nixpkgs/nixpkgs/pkgs/stdenv/generic/make-derivation.nix:380:7:
          379|       depsBuildBuild              = elemAt (elemAt dependencies 0) 0;
          380|       nativeBuildInputs           = elemAt (elemAt dependencies 0) 1;
             |       ^
          381|       depsBuildTarget             = elemAt (elemAt dependencies 0) 2;

       (stack trace truncated; use '--show-trace' to show the full, detailed trace)

       error: Package ‘dotnet-sdk-6.0.428’ in /nix/store/k5g5frkf231kzsrnc8szlg229ychpcqy-nixpkgs/nixpkgs/pkgs/development/compilers/dotnet/build-dotnet.nix:212 is marked as insecure, refusing to evaluate.


       Known issues:
        - Dotnet SDK 6.0.428 is EOL, please use 8.0 (LTS) or 9.0 (Current)

       You can install it anyway by allowing this package, using the
       following methods:

       a) To temporarily allow all insecure packages, you can use an environment
          variable for a single invocation of the nix tools:

            $ export NIXPKGS_ALLOW_INSECURE=1

          Note: When using `nix shell`, `nix build`, `nix develop`, etc with a flake,
                then pass `--impure` in order to allow use of environment variables.

       b) for `nixos-rebuild` you can add ‘dotnet-sdk-6.0.428’ to
          `nixpkgs.config.permittedInsecurePackages` in the configuration.nix,
          like so:

            {
              nixpkgs.config.permittedInsecurePackages = [
                "dotnet-sdk-6.0.428"
              ];
            }

       c) For `nix-env`, `nix-build`, `nix-shell` or any other Nix command you can add
          ‘dotnet-sdk-6.0.428’ to `permittedInsecurePackages` in
          ~/.config/nixpkgs/config.nix, like so:

            {
              permittedInsecurePackages = [
                "dotnet-sdk-6.0.428"
              ];
            }
```

</details>

## Additional context

I am running direnv with this shell.nix:
```
let
    pkgs = import <nixpkgs> {
        config = {
            allowUnfree = true;
            # azure-functions-core-tools currently relies on this
            permittedInsecurePackages = [
              "dotnet-sdk-wrapped-6.0.428"
            ];
        };
        overlays = [];
    };
in
with pkgs;
mkShellNoCC {
    name = "MitC";
    packages = [
        azure-cli
        # azure-functions-core-tools
        jq
        python312Full
        pipenv
        terraform
    ];
}
```

and the channels

```
nix-channel --list
home-manager https://github.com/nix-community/home-manager/archive/master.tar.gz
nixpkgs https://nixos.org/channels/nixpkgs-unstable

```

Strangely, adding the `permittedInsecurePackages` to the `shell.nix` made no difference?

The issue started to occur on 2024-11-28. I presume that updating the dotnet version in the package source should do the trick.

Thank you very much for all your efforts maintaining this! 😃 

## Metadata

 nix-shell -p nix-info --run "nix-info -m"
these 15 paths will be fetched (58.44 MiB download, 269.48 MiB unpacked):
  /nix/store/va7kw1i822h95im4jacci19v0cqajfyf-binutils-2.43.1
  /nix/store/mkgxzw8bdg0974njrhq29kgwmgk35wid-binutils-2.43.1-lib
  /nix/store/9bacph866qhmr2zfib1h49jixq5hhd01-binutils-wrapper-2.43.1
  /nix/store/swvs52wx5a5j9mv91mingbak5pn4h1i0-expand-response-params
  /nix/store/xzfmarrq8x8s4ivpya24rrndqsq2ndiz-gcc-13.3.0
  /nix/store/3ix5h74n7ar9950vwzp4dxmil70pmx0k-gcc-wrapper-13.3.0
  /nix/store/26dy2y29fhissya64ygahwpidpvs3bm3-glibc-2.40-36-bin
  /nix/store/aax0hx68i2ikhpf27hdm6a2a209d4s6p-glibc-2.40-36-dev
  /nix/store/617chz8a4p5kz09ij8s9ns05iafzbpcc-gmp-6.3.0
  /nix/store/zj93rq650fz6v25qw39sm025ya4k8plp-isl-0.20
  /nix/store/h82a48wb6i6g42fbpxl7z7s40pw93g0n-libmpc-1.3.1
  /nix/store/dvsai0ym9czjl5mcsarcdwccb70615n4-linux-headers-6.10
  /nix/store/93ph6f74wihfir4vyz2sj88d2vn2vq7y-mpfr-4.2.1
  /nix/store/6c7dwicl4jn20bg2zh0sbkpnlfpplsh8-nix-info
  /nix/store/gw0swsnbfk9l8d1il70psm8b3r5m33jv-stdenv-linux
copying path '/nix/store/6c7dwicl4jn20bg2zh0sbkpnlfpplsh8-nix-info' from 'https://cache.nixos.org'...
copying path '/nix/store/swvs52wx5a5j9mv91mingbak5pn4h1i0-expand-response-params' from 'https://cache.nixos.org'...
copying path '/nix/store/dvsai0ym9czjl5mcsarcdwccb70615n4-linux-headers-6.10' from 'https://cache.nixos.org'...
copying path '/nix/store/26dy2y29fhissya64ygahwpidpvs3bm3-glibc-2.40-36-bin' from 'https://cache.nixos.org'...
copying path '/nix/store/617chz8a4p5kz09ij8s9ns05iafzbpcc-gmp-6.3.0' from 'https://cache.nixos.org'...
copying path '/nix/store/mkgxzw8bdg0974njrhq29kgwmgk35wid-binutils-2.43.1-lib' from 'https://cache.nixos.org'...
copying path '/nix/store/zj93rq650fz6v25qw39sm025ya4k8plp-isl-0.20' from 'https://cache.nixos.org'...
copying path '/nix/store/93ph6f74wihfir4vyz2sj88d2vn2vq7y-mpfr-4.2.1' from 'https://cache.nixos.org'...
copying path '/nix/store/va7kw1i822h95im4jacci19v0cqajfyf-binutils-2.43.1' from 'https://cache.nixos.org'...
copying path '/nix/store/h82a48wb6i6g42fbpxl7z7s40pw93g0n-libmpc-1.3.1' from 'https://cache.nixos.org'...
copying path '/nix/store/aax0hx68i2ikhpf27hdm6a2a209d4s6p-glibc-2.40-36-dev' from 'https://cache.nixos.org'...
copying path '/nix/store/xzfmarrq8x8s4ivpya24rrndqsq2ndiz-gcc-13.3.0' from 'https://cache.nixos.org'...
copying path '/nix/store/9bacph866qhmr2zfib1h49jixq5hhd01-binutils-wrapper-2.43.1' from 'https://cache.nixos.org'...
copying path '/nix/store/3ix5h74n7ar9950vwzp4dxmil70pmx0k-gcc-wrapper-13.3.0' from 'https://cache.nixos.org'...
copying path '/nix/store/gw0swsnbfk9l8d1il70psm8b3r5m33jv-stdenv-linux' from 'https://cache.nixos.org'...
 - system: `"x86_64-linux"`
 - host os: `Linux 6.8.0-49-generic, Ubuntu, 24.04.1 LTS (Noble Numbat), nobuild`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.0pre20240627_b44909ac`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/home/michaela/.nix-defexpr/channels/nixpkgs`

## Notify maintainers

@mdarocha
@detegr
---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by Detegr**
Closes #360154

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by Detegr** [comment]
I'm getting the following error locally:
```
       > Patchelfing microsoft.azure.functions.dotnetisolatednativehost/1.0.11/contentFiles/any/any/workers/dotnet-isolated/bin/FunctionsNetHost.dbg as microsoft.azure.functions.dotnetisolatednativehost/1.0.11/contentFiles/any/any/workers/dotnet-isolated/bin/FunctionsNetHost.dbg.47.nix-patched
       > patchelf: patchelf.cc:1095: void setSubstr(std::string&, unsigned int, const std::string&): Assertion `pos + t.size() <= s.size()' failed.
```

Made the pull request to make it reproducible and/or to see if it happens in the CI as well.

**Comment by gepbird** [comment]
@NixOS/dotnet any idea about the aforementioned error?

**Comment by GGG-KILLER** [comment]
It seems like an issue with the way we detect what can be `patchelf`'d or not. I'll try to make a fix for it.

**Comment by GGG-KILLER** [comment]
@Detegr or @gepbird could either of you check if the changes in https://github.com/NixOS/nixpkgs/pull/361450 make it build properly for you now?

**Comment by gepbird** [comment]
> @Detegr or @gepbird could either of you check if the changes in #361450 make it build properly for you now?

@GGG-KILLER I cherry picked the commit from your PR on top of this and unfortunately the build failed.

<details><summary>Build log</summary>

```
       > Running phase: fixupPhase                                                                                                                                                                                
       > Patching package microsoft.azure.functions.dotnetisolatednativehost/1.0.11                                                                           
       > patchelf: cannot find section '.interp'. The input file is most likely statically linked                                                             
       > Patchelfing microsoft.azure.functions.dotnetisolatednativehost/1.0.11/contentFiles/any/any/workers/dotnet-isolated/bin/FunctionsNetHost as microsoft.azure.functions.dotnetisolatednativehost/1.0.11/contentFiles/any/any/workers/dotnet-isolated/bin/FunctionsNetHost.47.nix-patched
       > Patchelfing microsoft.azure.functions.dotnetisolatednativehost/1.0.11/contentFiles/any/any/workers/dotnet-isolated/bin/FunctionsNetHost.dbg as microsoft.azure.functions.dotnetisolatednativehost/1.0.11/contentFiles/any/any/workers/dotnet-isolated/bin/FunctionsNetHost.dbg.47.nix-patched
       > patchelf: patchelf.cc:1095: void setSubstr(std::string&, unsigned int, const std::string&): Assertion `pos + t.size() <= s.size()' failed.                                                                                                                                                                         
       > /nix/store/ng4s9mlk4phhnz4d59c14srfhrsv1cnz-patch-nupkgs/bin/patch-nupkgs: line 17:    72 Aborted                 (core dumped) /nix/store/xjlvb8x7z3f5rxxswa3c6ghay74a58jm-patchelf-0.15.0/bin/patchelf --add-needed libicui18n.so --add-needed libicuuc.so --add-needed libz.so --add-needed libssl.so "$tmp"
```
</details>

**Comment by GGG-KILLER** [comment]
Ok, got it to compile with the latest changes in #361450, but now there's an error on the `buildDotnetModule` usage:
```
@nix { "action": "setPhase", "phase": "installPhase" }
Executing dotnetInstallHook
/nix/store/wlknwwx8f79lq8n07b0xh7ddjrrk191h-dotnet-sdk-8.0.404/share/dotnet/sdk/8.0.404/Current/SolutionFile/ImportAfter/Microsoft.NET.Sdk.Solution.targets(36,5): warning NETSDK1194: The "--output" option isn't supported when building a solution. Specifying a solution-level output path results in all projects copying outputs to the same directory, which can lead to inconsistent builds. [/build/source/Azure.Functions.Cli.sln]
/nix/store/wlknwwx8f79lq8n07b0xh7ddjrrk191h-dotnet-sdk-8.0.404/share/dotnet/sdk/8.0.404/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.CrossTargeting.targets(31,5): error NETSDK1129: The 'Publish' target is not supported without specifying a target framework. The current project targets multiple frameworks, you must specify one of the following frameworks in order to publish: net8.0 [/build/source/test/Azure.Functions.Cli.Tests/Azure.Functions.Cli.Tests.csproj]
/nix/store/wlknwwx8f79lq8n07b0xh7ddjrrk191h-dotnet-sdk-8.0.404/share/dotnet/sdk/8.0.404/Sdks/Microsoft.NET.Sdk/targets/Microsoft.NET.Sdk.CrossTargeting.targets(31,5): error NETSDK1129: The 'Publish' target is not supported without specifying a target framework. The current project targets multiple frameworks, you must specify one of the following frameworks in order to publish: net8.0 [/build/source/src/Azure.Functions.Cli/Azure.Functions.Cli.csproj]
```

**Comment by GGG-KILLER** [comment]
PR fixing the bug you were encountering has been merged, if you could rebase and check it it works.

**Comment by Detegr** [comment]
Updated the PR with `dotnetFlags` set. Seems to be working now, thanks!

**Comment by corngood** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 361077`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>azure-functions-core-tools</li>
  </ul>
</details>

**Review by corngood — APPROVED** [review]

**Review by gepbird — APPROVED** [review]
Thanks, your changes LGTM and the package works with .NET 8!

A few improvements and nits, but feel free to ignore if you think it's too unrealted:
- run nixfmt on the package.nix
- drop `--replace` as it is deprecated
```diff
-      --replace "CheckExitCode(\"/bin/bash" "CheckExitCode(\"${stdenv.shell}"
+      --replace-fail "CheckExitCode(\"/bin/bash" "CheckExitCode(\"${stdenv.shell}"
```
- remove `with lib;` from `meta`
- version checking
```diff

+  doInstallCheck = true;
+  nativeInstallCheckInputs = [
+    versionCheckHook
+  ];
  ```
- set `meta.mainProgram`
- add an update script
- remove `rec` next to `buildDotnetModule`

**Comment by GGG-KILLER** [comment]
I wouldn't ignore the nixfmt point nor the `--replace-fail` one. They're both important.

**Comment by corngood** [comment]
> I wouldn't ignore the nixfmt point nor the `--replace-fail` one. They're both important.
> 
> The others would be _tolerable_ if not done (but should be done ideally), however without `meta.mainProgram`, `lib.getExe` and `nix run` may not work properly. But none of them are unrelated overall IMHO.

This is all good stuff, but this is essentially a simple update PR, and I don't think it makes sense to block it.

**Comment by Detegr** [comment]
Fixed review comments. I had to left out version check as the software seems to report `4.0.1` for every `4.0.xxxx` version :shrug:

**Review by corngood — APPROVED** [review]

**Comment by corngood** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 361077`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 1 package built:</summary>
  <ul>
    <li>azure-functions-core-tools</li>
  </ul>
</details>

File: pkgs/by-name/az/azure-functions-core-tools/package.nix [hunk 1]
```diff
@@ -1,17 +1,19 @@
-{ lib,
+{
+  lib,
   stdenv,
   fetchFromGitHub,
   buildDotnetModule,
   buildGoModule,
   dotnetCorePackages,
+  versionCheckHook,
```
Comment (gepbird) [inline pkgs/by-name/az/azure-functions-core-tools/package.nix:8]: Unused, but its not worth blocking on this. Unfortunate that upstream doesn't properly update their version
```suggestion
```

**Review by gepbird — APPROVED** [review]

**Pull request merged**